HoRNDIS.cpp

```diff
@@ -891,74 +858,80 @@ IOReturn HoRNDIS::message(UInt32 type, IOService *provider, void *argument) {
 		LOG(V_NOTE, "kIOMessageServiceIsAttemptingOpen");
 		break;
 	default:
+		LOG(V_ERROR, ">>>>>>>>> Possibly un-updated messages!!!");
 		LOG(V_NOTE, "unknown message type %08x", (unsigned int) type);
 		break;
 	}
-	
-	return kIOReturnUnsupported;
+
+	LOG(V_ERROR, "###################### Received the message: %d, ", type);
+	return super::message(type, provider, argument);
 }
+*/
 
 
 /***** Packet transmit logic *****/
 
+static inline bool isTransferStopStatus(IOReturn rc) {
+	// IOReturn indicating that we need to stop transfers:
+	return rc == kIOReturnAborted || rc == kIOReturnNotResponding;
+}
+
 UInt32 HoRNDIS::outputPacket(mbuf_t packet, void *param) {
-	mbuf_t m;
-	size_t pktlen = 0;
 	IOReturn ior = kIOReturnSuccess;
-	UInt32 poolIndx;
-	int i;
-
-	LOG(V_DEBUG, "");
+	int poolIndx = N_OUT_BUFS;
+
+	// Note, this function MAY or MAY NOT be protected by the IOCommandGate,
+	// depending on the kind of OutputQueue used.
+	// Here, we assume that IOCommandGate is used: no need to lock.
+
+	if (!fReadyToTransfer) {
+		// Technically, we must never be here, because we always disable the
+		// queue before clearing 'fReadyToTransfer', but double-checking here
+		// just in-case: better safe than sorry.
+		LOG(V_DEBUG, "fReadyToTransfer=false: dropping packet "
+			"(we shouldn't even be here)");
+		freePacket(packet);
+		return kIOReturnOutputDropped;
+	}
 	
-	/* Count the total size of this packet */
-	m = packet;
-	while (m) {
+	// Count the total size of this packet
+	size_t pktlen = 0;
+	for (mbuf_t m = packet; m; m = mbuf_next(m)) {
 		pktlen += mbuf_len(m);
-		m = mbuf_next(m);
 	}
 	
-	LOG(V_DEBUG, "%ld bytes", pktlen);
+	LOG(V_PACKET, "%ld bytes", pktlen);
+
+	const uint32_t transmitLength = (uint32_t)(pktlen + sizeof(rndis_data_hdr));
 	
-	if (pktlen > (mtu + 14)) {
-		LOG(V_ERROR, "packet too large (%ld bytes, but I told you you could have %d!)", pktlen, mtu);
+	if (transmitLength > maxOutTransferSize) {
+		LOG(V_ERROR, "packet too large (%ld bytes, maximum can transmit %ld)",
+			pktlen, maxOutTransferSize - sizeof(rndis_data_hdr));
 		fpNetStats->outputErrors++;
-		return false;
+		freePacket(packet);
+		return kIOReturnOutputDropped;
 	}
-	
-	/* Find an output buffer in the pool */
-	IOLockLock(outbuf_lock);
-	for (i = 0; i < OUT_BUF_MAX_TRIES; i++) {
-		uint64_t ivl, deadl;
-		
-		for (poolIndx = 0; poolIndx < N_OUT_BUFS; poolIndx++)
-			if (!outbufs[poolIndx].inuse) {
-				outbufs[poolIndx].inuse = true;
-				break;
-			}
-		if (poolIndx != N_OUT_BUFS)
-			break;
-		
-		/* "while", not "if".  See Symphony X's seminal work on this topic, /Paradise Lost/ (2007). */
-		nanoseconds_to_absolutetime(OUT_BUF_WAIT_TIME, &ivl);
-		clock_absolutetime_interval_to_deadline(ivl, &deadl);
-		LOG(V_NOTE, "waiting for buffer...");
```

Comment:
This kind of thing actually used to happen some times, by the way -- the USB stack would temporarily lose its mind.  So being able to time things out eventually might be good.

Reply:
Well, it's true that you may want to time out eventually, but timing out here would be of zero benefit towards getting more free buffers. 
My answer, if we get stuck for a long time (for some unknown reason), the best course of action would be for the user to disable tethering (or unplug the device), and then re-enable tethering (plug in). When we unplug, all the pending reads/writes should abort, and device should be cleanly freed.